### ISSUE TITLE
Unicode View Support For OutputDebugStringW 

### ISSUE THREAD

**Issue by luodaoyi** [CobaltFusion/DebugViewPP#389]
https://learn.microsoft.com/en-us/windows/win32/api/debugapi/nf-debugapi-waitfordebugeventex

https://learn.microsoft.com/zh-cn/windows/win32/api/debugapi/nf-debugapi-outputdebugstringw#remarks

OutputDebugStringW does internally call OutputDebugStringA, so Unicode characters that cannot be represented in the system code page will be replaced with ?.

Oddly enough, the [OUTPUT_DEBUG_STRING_INFO](http://msdn.microsoft.com/en-us/library/windows/desktop/ms680545%28v=vs.85%29.aspx) structure the debugger receives from the operating system to print the message does appear to support letting the debugger know if the string is Unicode, it just doesn't appear to be used by OutputDebugStringW at all.

Unfortunately, I don't know of a mechanism to get the OS to raise a [OUTPUT_DEBUG_STRING_EVENT](http://msdn.microsoft.com/en-us/library/windows/desktop/ms679308%28v=vs.85%29.aspx) with a Unicode string. It may not be possible with public APIs.

ref:  

 [WaitForDebugEventEx function (debugapi.h)](https://learn.microsoft.com/en-us/windows/win32/api/debugapi/nf-debugapi-waitfordebugeventex)

 [windows - Is there a Unicode alternative to OutputDebugString? - Stack Overflow](https://stackoverflow.com/questions/19442535/is-there-a-unicode-alternative-to-outputdebugstring)

**Comment by janwilmans** [comment]
I am not sure what the question is here? if you are looking for a way to
send special characters, I suggest you try to use UTF-8 encoding through
the  OutputDebugStringA function? I think this how chinese users of
debugview are doing it...

Greetings,

Jan

**Comment by luodaoyi** [comment]
Example:

Change setting to Japanese. or Korean language.
Then call OutputDebugStringW("中文") with  Chinese  unicode charset.

In DebugView++ show: "????";

**Comment by janwilmans** [comment]
![image](https://github.com/CobaltFusion/DebugViewPP/assets/5933444/0ec75876-aea3-4a13-9fb4-aff32fc9a915)

**Comment by janwilmans** [comment]
Please try changing the setting above, you can find it by searching for "Region" in the start menu

**Comment by janwilmans** [comment]
> Example:
> 
> Change setting to Japanese. or Korean language. Then call OutputDebugStringW("中文") with Chinese unicode charset.
> 
> In DebugView++ show: "????";

You mean when the language in the OutputDebugStringW is different from the non-unicode programs setting in the Region dialog?

**Comment by WebDancer69** [comment]
> You mean when the language in the OutputDebugStringW is different from the non-unicode programs setting in the Region dialog?

Not necessarily the "language", but a lot of characters does not exist in the (limited) code page.
Like e.g. Greek characters (in math) and emojis. Even if you try to avoid using "foreign" / "complex" characters in debug messages you often print project data, which can contain any Unicode glyph.

> Later>> no sorry I was confusing two projects, debugviewpp has no UTF-8 support.

Are you sure DebugViewPP does not support it? It should support it if it supports other MBCS.
![image](https://github.com/CobaltFusion/DebugViewPP/assets/9715604/7efe6054-2098-4cb3-bed6-3c15ed7873e4)

**Comment by janwilmans** [comment]
well, debugview does not support UTF-8 natively, but this windows feature might help, I am not sure. I will do some more testing. Let me know if you already found a solution?

**Comment by janwilmans** [comment]
2026 update: 

As mentioned in the first message,  by default Windows /  OutputDebugStringW internally calls OutputDebugStringA, so for example:

```
void EncodingTest()
{
    const wchar_t* utf16Message = L"This message is UTF-16 encoded and send through OutputDebugStringW \u65E5\u672C\u8A9E"; // 日本語 UTF-16 encoded
    OutputDebugStringW(utf16Message);

    // 日本語 😀" encoded as UTF-8 bytes
    const char* utf8Message = "This message is UTF-8 encoded and send through OutputDebugStringA (with special windows settings): "
        "\xE6\x97\xA5" // 日 // Ni
        "\xE6\x9C\xAC" // 本 // Hon
        "\xE8\xAA\x9E" // 語 // Go
        " "
        "\xF0\x9F\x98\x80"; // 😀

    OutputDebugStringA(utf8Message);
}
```

The code about will look like this (by default) in DebugView++

<img width="1304" height="74" alt="Image" src="https://github.com/user-attachments/assets/36e0e997-890a-410d-8f56-d17191cc4afc" />

You can see that 日本語 passed through OutputDebugStringW is literally translated into a call `OutputDebugStringA("???");` by windows, so Debugview++ receives the literal text '???' and can't do anything for WStrings.

Similarly, but with a slightly different result, the UTF-8 string is passing through just fine, but the font rendering threats the string as ASCII. 

# solution / workaround

As of windows 11 the UTF-8 'feature' for non-unicode programs is still in 'beta'. 

<img width="883" height="294" alt="Image" src="https://github.com/user-attachments/assets/4d61ff23-e5d9-427c-bcd7-559f46d03268" />

By selecting this option: `"Beta: Use Unicode UTF-8 for worldwide language support: ON"`

Selecting this option has two effects:
- OutputDebugStringW will not 'translate' strings into ASCII ??? characters anymore
- The font renderer will accept UTF-16 and UTF-8 encoded strings 

As a result you will see Debugview++ now displayed both correctly. (But in reality Debugview++ didn't do anything, it only passed through the string as-is to windows font rendering.

The result is properly displayed Unicode strings for both cases:

<img width="1245" height="99" alt="Image" src="https://github.com/user-attachments/assets/c3792bb4-0445-467d-bc58-527d23afc5e5" />

**Comment by janwilmans** [comment]
With this explanation and https://github.com/CobaltFusion/DebugViewPP/pull/408 all should be working now

**Comment by WebDancer69** [comment]
You can also make a given app always use UTF-8, instead of having to set that globally. Did you test that?
You can do this in the application manifest using the activeCodePage element:
https://learn.microsoft.com/en-us/windows/win32/sbscs/application-manifests#activecodepage

**Comment by janwilmans** [comment]
That sounds like it could be very useful, any idea how to do that from a C++ ATL project? Maybe you could create a pull-request?

**Comment by janwilmans** [comment]
Hmm chat-gpt tells me I should write an `app.manifest` file and cmake will embed it in the executable during the linking step if I reference it as 
```
target_sources(MyAtlApp PRIVATE app.manifest)
```

I will have to try that ;)

```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0">

  <application>
    <windowsSettings xmlns="http://schemas.microsoft.com/SMI/2016/WindowsSettings">
      <activeCodePage xmlns="http://schemas.microsoft.com/SMI/2019/WindowsSettings">
        UTF-8
      </activeCodePage>
    </windowsSettings>
  </application>

</assembly>
```

**Comment by WebDancer69** [comment]
Yes, easy to do. But what happens if the sender app and DebugViev++ use different codepages? I.e. both does not use UTF-8.

**Comment by janwilmans** [comment]
well, I can test that but I can imagine a couple of scenarios, either we build a normal and a UTF-8 version and distribute both, or we can say version 1.3.x is just UTF-8 and if want Latin-1, use an older version of Debugview++ ;)
